examples/eip712.rs

```diff
@@ -0,0 +1,125 @@
+use std::{convert::TryFrom, sync::Arc, time::Duration};
+
+use ethers::{
+    contract::EthAbiType,
+    prelude::*,
+    types::{transaction::eip712::Eip712, Address, I256, U256},
+    utils::{compile_and_launch_ganache, keccak256, Ganache, Solc},
+};
+use ethers_derive_eip712::*;
+
+#[derive(Debug, Clone, Eip712, EthAbiType)]
+#[eip712(
+    name = "Eip712Test",
+    version = "1",
+    chain_id = 1,
+    verifying_contract = "0x0000000000000000000000000000000000000001",
```

Comment:
Do you think it'd be worth allowing functions which specify `chain_id` and `verifying_contract` at runtime? That way if Im testing on both mainnet and testnet, I don't need to modify my code and can just provide it as a param.

Reply:
Perhaps we can make the `EIP712Domain` type part of the trait so that instead of the hash, `domain_type` returns `EIP712Domain` which has `EIP712Domain::encode` and additional setters. This way we can add a type that wraps around a `EIP712Domain` and takes an `Eip712` type as reference when encoding the actual eip712.

Reply:
I like the flexible to pass in runtime values for the domain. After considering the above solution, I opted for providing an extra parameter for the domain when calling encode: 

```rust
fn encode_eip712(self, domain: Option<EIP712Domain>) -> Result<[u8; 32], Self::Error>;
```

I found this to be a simple solution, without needing to implement a Eip712 type, which may not actually correspond to the domain type hash. I think using the `EIP712Domain` type is more defensible against user error.

The macro implementation looks like:

```rust
fn encode_eip712(self, domain: Option<ethers_core::types::transaction::eip712::EIP712Domain>) -> Result<[u8; 32], Self::Error> {
                // encode the digest to be compatible with solidity abi.encodePacked()
                // See: https://github.com/gakonst/ethers-rs/blob/master/examples/permit_hash.rs#L72

                let domain_separator = if let Some(d) = domain {
                    d.separator()
                } else {
                    Self::domain_separator()?
                };


                let digest_input = [
                    &[0x19, 0x01],
                    &domain_separator[..],
                    &self.struct_hash()?[..]
                ].concat();

                return Ok(ethers_core::utils::keccak256(digest_input));
}
```

This also means we need to expose the parameter to the signer trait for `sign_typed_data`, like:

```rust
/// Encodes and signs the typed data according EIP-712.
    /// Payload must implement Eip712 trait.
    /// * `domain` - Optional Eip712 domain struct to override eip712 macro attribute helpers for Eip712 Type `T`;
    async fn sign_typed_data<T: Eip712 + Send + Sync>(
        &self,
        payload: T,
        domain: Option<EIP712Domain>,
    ) -> Result<Signature, Self::Error>;
```

Overall, I think this might be quick solution to the request, but has the downside of needed an extra parameter, but I don't think that is too large a burden.

Open to other feedback here, but we probably should look to open a new PR to introduce such a request and get the current works landed.